### PR TITLE
feat: implement AuthPolicyMatcher interface for LLM service  discovery

### PR DESCRIPTION
**Pull request by Jooho**
## Description

  - Introduce new `AuthPolicyMatcher` interface for discovering LLMInferenceService from AuthPolicy resources
  - Implement `KServeAuthPolicyMatcher` with support for both HTTPRoute and Gateway AuthPolicy matching
  - Add comprehensive test suite covering various matching scenarios and edge cases

  ## Changes
  - **New interface**: `AuthPolicyMatcher` with methods for HTTPRoute and Gateway AuthPolicy matching
  - **Implementation**: `KServeAuthPolicyMatcher` that finds LLM services via:
    - OwnerReference lookup (primary method)
    - HTTPRoute name pattern matching (fallback)
    - Gateway-based matching with LLMInferenceService enumeration
  - **Tests**: Full test coverage for both positive and negative scenarios

https://issues.redhat.com/browse/RHOAIENG-34436

## How Has This Been Tested?

**Not needed**
- Configuration-only changes: Cache filtering and label additions are infrastructure setup
  using well-tested Kubernetes/controller-runtime features, not new business logic requiring
  validation.

 - Existing coverage sufficient: AuthPolicy creation and management logic remains unchanged and
   is already covered by existing tests.
## Merge criteria:

- [x] JIRA(s) are linked in the PR description
- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: internal/controller/resources/authpolicy.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/go-logr/logr"
 	kservev1alpha1 "github.com/kserve/kserve/pkg/apis/serving/v1alpha1"
 	kuadrantv1 "github.com/kuadrant/kuadrant-operator/api/v1"
+	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
```
[the remark this answers is not in the document]
Comment (Jooho) [reply internal/controller/resources/authpolicy.go]: updated

File: internal/controller/resources/authpolicy.go [hunk 2]
```diff
@@ -273,3 +279,82 @@ func (c *clientAuthPolicyStore) Update(ctx context.Context, authPolicy *kuadrant
...
+		return authPolicy.Namespace == gatewayNamespace && string(authPolicy.Spec.TargetRef.Name) == gatewayName
+	}
+
+	for _, ref := range llmSvc.Spec.Router.Gateway.Refs {
+		if string(ref.Name) == string(authPolicy.Spec.TargetRef.Name) && string(ref.Namespace) == authPolicy.Namespace {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (Jooho) [reply internal/controller/resources/authpolicy.go:360]: updated

File: internal/controller/serving/llm/llm_inferenceservice_controller.go [hunk 3]
```diff
@@ -94,16 +103,104 @@ func (r *LLMInferenceServiceReconciler) SetupWithManager(mgr ctrl.Manager, setup
...
+				return []reconcile.Request{{NamespacedName: namespacedName}}
+			}
+		}
+
+		if namespacedName, found := r.authPolicyMatcher.FindLLMServiceFromGatewayAuthPolicy(ctx, authPolicy, r.Client); found {
+			return []reconcile.Request{{NamespacedName: namespacedName}}
+		}
+		return []reconcile.Request{}
+	})
+}
```
[the remark this answers is not in the document]
Comment (Jooho) [reply internal/controller/serving/llm/llm_inferenceservice_controller.go:141]: I don’t think it’s possible for other types of objects to appear, since they are captured by the watcher.

File: internal/controller/resources/authpolicy.go [hunk 4]
```diff
@@ -273,3 +279,82 @@ func (c *clientAuthPolicyStore) Update(ctx context.Context, authPolicy *kuadrant
...
+		continueToken = llmSvcList.Continue
+	}
+
+	return types.NamespacedName{}, false
+}
+
+func (k *kserveAuthPolicyMatcher) isGatewayMatchedWithInfo(llmSvc *kservev1alpha1.LLMInferenceService, authPolicy *kuadrantv1.AuthPolicy, gatewayNamespace, gatewayName string) bool {
+	if llmSvc.Spec.Router == nil || llmSvc.Spec.Router.Gateway == nil || !llmSvc.Spec.Router.Gateway.HasRefs() {
+		return authPolicy.Namespace == gatewayNamespace && string(authPolicy.Spec.TargetRef.Name) == gatewayName
+	}
```
[the remark this answers is not in the document]
Comment (Jooho) [reply internal/controller/resources/authpolicy.go:352]: will be solved with authPolicy reconciler

File: internal/controller/resources/authpolicy.go [hunk 5]
```diff
@@ -273,3 +279,82 @@ func (c *clientAuthPolicyStore) Update(ctx context.Context, authPolicy *kuadrant
...
+	continueToken := ""
+	for {
+		llmSvcList := &kservev1alpha1.LLMInferenceServiceList{}
+		if err := c.List(ctx, llmSvcList, &client.ListOptions{Namespace: listNamespace, Continue: continueToken}); err != nil {
+			return types.NamespacedName{}, false
+		}
+
+		for _, llmSvc := range llmSvcList.Items {
+			if k.isGatewayMatchedWithInfo(&llmSvc, authPolicy, gatewayNamespace, gatewayName) {
+				return types.NamespacedName{
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply internal/controller/resources/authpolicy.go:333]: I think it's for future proofing where we might need to do more when such an event occurs. In my eyes we shouldn't stop on first match as it shouldn't be expensive to trigger reconcile for all instances.

**Review by bartoszmajsak — APPROVED** [review]
/lgtm
/approve
